LPDDR5 IO power calculation

alexcani opened the pull request.

alexcani committed:
Refactor some code style on LPDDR5

alexcani committed:
LPDDR5 command patterns

alexcani committed:
Add clocks to class

alexcani committed:
Handle interface commands on LPDDR5 (RD and WR) to calculate simulate DQ buses, WCK and DQS

alexcani committed:
IO power calculation LPDDR5 initial implementation

alexcani committed:
Instantiate IO power calc on LPDDR5

alexcani committed:
Misc changes

Added default return value for parseBoolWithDefault
Fixed scope issues with length variable in handle_interface methods
Made calc_static_power and calc_dynamic_power static (internal linkage)

alexcani committed:
Parse LPDDR5 impedance specs and bus rate specs

lsteiner-tukl commented on src/DRAMPower/DRAMPower/data/stats.h:
```diff
@@ -42,6 +42,7 @@ namespace DRAMPower
 		util::bus_stats_t readBus;
 		util::bus_stats_t writeBus;
 		util::bus_stats_t clockStats;
+		util::bus_stats_t WClockStats;
```
From our naming conventions lowercase would be better.

alexcani replied:
allright!

lsteiner-tukl commented on src/DRAMPower/DRAMPower/memspec/MemSpecLPDDR5.cpp:
```diff
@@ -1,10 +1,8 @@
-#include "MemSpecLPDDR5.h"
-
+#include "DRAMPower/memspec/MemSpecLPDDR5.h"
```
This change is not really necessary. We always place the corresponding header file in the same directory.

lsteiner-tukl commented on src/DRAMPower/DRAMPower/memspec/MemSpecLPDDR5.cpp:
```diff
@@ -95,6 +98,41 @@ uint64_t MemSpecLPDDR5::timeToCompletion(DRAMPower::CmdType type)
+    memImpedanceSpec.R_eq_wb = parseUdouble(memspec["memimpedancespec"]["R_eq_wb"], "R_eq_wb");
+}
 
+void MemSpecLPDDR5::parseDataRateSpec(nlohmann::json &memspec) {
```
Is this spec required? The rate between CK and data transmission rate (2*WCK) is specified as dataRate in the memarchitecturespec. And both DQ as well as DQS are running at those rates. Or is there a mode where they can have different rates?

alexcani replied:
I don`t think so. I removed the spec, since I copied it over from LPDDR4

lsteiner-tukl commented on src/DRAMPower/DRAMPower/standards/lpddr5/LPDDR5.cpp:
```diff
@@ -11,7 +11,7 @@ namespace DRAMPower {
 		, ranks(memSpec.numberOfRanks, { (std::size_t)memSpec.numberOfBanks })
-		, commandBus{6}
+		, commandBus{7}
 		, readBus{6}
```
Why are readBus and writeBus initialized with 6?

alexcani replied:
No idea, fixed it do 16

lsteiner-tukl commented on src/DRAMPower/DRAMPower/standards/lpddr5/LPDDR5.cpp:
```diff
@@ -38,65 +38,140 @@ namespace DRAMPower {
+                readDQS_t_.stop(cmd.timestamp + length / this->memSpec.dataRateSpec.dqsBusRate);
+
+                // WCK also during reads
+                if (!memSpec.wckAlwaysOnMode) {
```
Where do you add the power for WCK when we are in wckAlwaysOnMode?

alexcani replied:
The power is calculated based on the clock stats in getWindowStats, however there was a bug where the clocks weren't initialized properly. The DQS should be initialized stopped, and the wCK must be initialized depending on the wckAlwaysOnMode mode

lsteiner-tukl replied:
Ok, and when wckAlwaysOnMode is enabled the last_start is initialized with 0 and we always count the number of cycles between the current time and the start of the simulation, right?

alexcani replied:
yes, correct. that is done in the getWindowStats function

lsteiner-tukl commented on src/DRAMPower/DRAMPower/standards/lpddr5/LPDDR5.h:
Did you just rearrange the code in this file or are there some changes?

alexcani replied:
Only rearranging, there were way too many public, protected and private sections.
Only addition was adding    
```
    util::Clock WCK_t_;
    util::Clock WCK_c_;
    util::Clock readDQS_c_;
    util::Clock readDQS_t_;
```

lsteiner-tukl commented on src/DRAMPower/DRAMPower/standards/lpddr5/interface_calculation_LPDDR5.cpp:
Check termination scheme of LPDDR5 pins/signals.

alexcani replied:
Seems to be all right, did you notice anything wrong?

Commit:
Pull request feedback changes

Lowercase wClockStats
Remove DataRateSpec from LPDDR5
Initialize WCK and DQS in correct state
DQ bus width to 16

Commit:
Fix LPDDR5 core tests (initialize WCKtoCK)

lsteiner-tukl commented on src/DRAMPower/DRAMPower/standards/lpddr5/LPDDR5.cpp:
```diff
@@ -38,65 +42,140 @@ namespace DRAMPower {
     void LPDDR5::registerPatterns() {
         using namespace pattern_descriptor;
-
+        // All commands consider 16B bank architecture mode
```
We need to think of a way to extend this for BG mode.
